libvirt/resource_libvirt_domain.go

```diff
@@ -244,28 +248,83 @@ func resourceLibvirtDomainCreate(d *schema.ResourceData, meta interface{}) error
 		return fmt.Errorf("Error retrieving host architecture: %s", err)
 	}
 
+	// Setup graphics and video
 	if arch == "s390x" || arch == "ppc64" {
 		domainDef.Devices.Graphics = nil
 	} else {
 		if graphics, ok := d.GetOk("graphics"); ok {
-			graphics_map := graphics.(map[string]interface{})
+			graphicsMap := graphics.(map[string]interface{})
 			domainDef.Devices.Graphics = []libvirtxml.DomainGraphic{
 				libvirtxml.DomainGraphic{},
 			}
-			if graphics_type, ok := graphics_map["type"]; ok {
-				domainDef.Devices.Graphics[0].Type = graphics_type.(string)
+
+			if graphicsType, ok := graphicsMap["type"].(string); ok {
+				if !(graphicsType == "vnc" || graphicsType == "spice") {
+					return fmt.Errorf("[ERROR] We only support graphics types of spice or vnc, not: %s", graphicsType)
+				}
+				domainDef.Devices.Graphics[0].Type = graphicsType
+			} else {
+				domainDef.Devices.Graphics[0].Type = "spice"
+				d.Set("graphics.type", "spice")
 			}
-			if autoport, ok := graphics_map["autoport"]; ok {
-				domainDef.Devices.Graphics[0].AutoPort = autoport.(string)
+
+			// If both autoport is yes and port are set then we should fail
+			// If neither are set use autoport
+			// Otherwise use autoport or port as specified
+			if graphicsMap["autoport"] != nil && graphicsMap["listen_port"] != nil {
+				return fmt.Errorf("[ERROR] It does not make sense to set autoport to yes and a listen_port")
+			} else if (graphicsMap["autoport"] == nil && graphicsMap["listen_port"] == nil) ||
+				(graphicsMap["autoport"] != nil && graphicsMap["autoport"].(string) == "yes") {
+				domainDef.Devices.Graphics[0].AutoPort = "yes"
+			} else {
+				log.Printf("[INFO] As port has been specified turning autoport off")
+				domainDef.Devices.Graphics[0].AutoPort = "no"
+				d.Set("graphics.autoport", "no")
+				domainDef.Devices.Graphics[0].Port, _ = strconv.Atoi(graphicsMap["listen_port"].(string))
 			}
```

Comment:
IMO this section looks unnecessary complicated, and doesn't cover all cases. For example, this code will fail if `listen_port != nil` and `autoport == "no"`. This shouldn't be, and there's no test case for it.

 Also, go can [check the existence of keys in maps](https://tour.golang.org/moretypes/22). There's no need to check for `nil`. This should be simpler and cover all cases:
```go
autoPort, autoPortOk := graphicsMap["autoport"]
listenPort, listenPortOk := graphicsMap["listen_port"]

// If both autoport is yes and port are set then we should fail
if autoPort == "yes" && listenPortOk {
	return fmt.Errorf("[ERROR] autoport and listen_port cannot be used at the same time")
}

// If neither are set use autoport
if !autoPortOk && !listenPortOk {
	domainDef.Devices.Graphics[0].AutoPort = "yes"
} else {
	// Otherwise use autoport and/or port as specified
	if autoPortOk {
		domainDef.Devices.Graphics[0].AutoPort = autoPort.(string)
	}

	if listenPortOk {
		domainDef.Devices.Graphics[0].Port, _ = strconv.Atoi(listenPort.(string))
	}
}
```